etc/completions/flux-core

```diff
@@ -0,0 +1,45 @@
+ #
+ # Copyright 2014 Lawrence Livermore National Security, LLC
+ # (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ #
+ # This file is part of the Flux resource manager framework.
+ # For details, see https://github.com/flux-framework.
+ #
+ # SPDX-License-Identifier: LGPL-3.0
+ #
+
+_flux_core() 
+{
+    local cur prev longopts comopts
+    COMPREPLY=()
+    cur="${COMP_WORDS[COMP_CWORD]}"
+    prev="${COMP_WORDS[COMP_CWORD-1]}"
+
+    longopts="--parent --verbose --version"
+    comopts="broker content cron dmesg env event exec get \
+             set lsattr hwloc jobs keygen kvs logger mini \
+             module ping proxy start user version help"
```

Comment:
We should try to generate this list of subcommands automatically. There are a couple solutions that come to mind, feel free to try a few things out

 * run a `find` on `FLUX_EXEC_PATH` to get a list of all executables that start with `flux-`, e.g.
```
 (IFS=:; for f in $FLUX_EXEC_PATH; do find $f -maxdepth 1 -type f -executable -name flux-\* ; done)
/g/g0/grondo/git/f.git/src/broker/flux-broker
/g/g0/grondo/git/f.git/src/cmd/flux-cron
/g/g0/grondo/git/f.git/src/cmd/flux-jobspec.py
/g/g0/grondo/git/f.git/src/cmd/flux-mini.py
/g/g0/grondo/git/f.git/src/cmd/flux-keygen
/g/g0/grondo/git/f.git/src/cmd/flux-event
/g/g0/grondo/git/f.git/src/cmd/flux-logger
/g/g0/grondo/git/f.git/src/cmd/flux-aggregate
/g/g0/grondo/git/f.git/src/cmd/flux-comms
/g/g0/grondo/git/f.git/src/cmd/flux-module
/g/g0/grondo/git/f.git/src/cmd/flux-ping
/g/g0/grondo/git/f.git/src/cmd/flux-queue
/g/g0/grondo/git/f.git/src/cmd/flux-exec
/g/g0/grondo/git/f.git/src/cmd/flux-start
/g/g0/grondo/git/f.git/src/cmd/flux-kvs
/g/g0/grondo/git/f.git/src/cmd/flux-job
/g/g0/grondo/git/f.git/src/cmd/flux-jobs.py
```
(you'll have to strip the leading `*flux-` and any trailing `.py`).
This has the benefit of including site-installed subcommands, or subcommands that were added by flux-framework projects.

This won't list builtin commands though, so those will have to be added at build time (there's a script `src/cmd/builtin-cmds-list.sh` that could be modified to get the builtins though)

 * A second approach would be to make the `flux(1)` cmd driver itself emit the list of possible subcommands. A hidden  builtin command or an option could be added which would print all possible subcommands given the current environment and builtin paths. Eventually `optparse` itself and Python argparse could be extended with similar options, so that commands at each level could list subcommands (and options) in a standard way (I think something like this was already suggested by @SteVwonder)

Reply:
Thanks for the suggestions on generating the subcommands dynamically! 

However, when I generate the list of subcommands and builtins, I get some discrepancies between the generated commands and the commands listed by flux help. 

The list of subcommands (first row) and builtins (second row) generated:
<img width="824" alt="image" src="https://user-images.githubusercontent.com/37486957/77836216-7cc6af80-7111-11ea-816e-a111edfe3ced.png">

The list of commands from flux help:
<img width="630" alt="image" src="https://user-images.githubusercontent.com/37486957/77836324-8c92c380-7112-11ea-8dbb-a6c2ab284ca0.png">

For example, from the subcommands, aggregate, jobspec, job, queue, and comms are all not listed under flux help. Does anyone have any suggestions/things I may be missing?

Reply:
Very cool that you got the auto-generation working!

I think it is intended that some commands be "blacklisted" from the `flux --help` listing.  I'm surprised though that `job` and `queue` aren't already in the `--help` listing. Hmmmm.

Maybe it is a separate issue as to which commands exactly are blacklisted, but I think as long as the blacklist functionality exists, we can tune and tweak it later.

Reply:
`flux help` reads the list of commands from `$(datadir)/flux/help.d/*.json`. For flux-core, the JSON input is autogenerated by `etc/gen-cmdhelp.pl`, which looks through the `*.adoc` manpages looking for a `flux-help-include: true` tag. This way, as commands are added/removed they automatically appear/disappear from `flux help`. The drawback is if there is no manpage, then the command won't appear in `flux-help`.

We'll probably want to expand this system to have different "levels" of help in the near future (perhaps displaying different results for owner vs guest if possible.)